Optimise hail size plugin

MoseleyS opened the pull request:
As part of the hail chain testing, we found that the hail size CLI was very slow and memory hungry. This PR attempts to address that by:

- Changing the heart of the plugin to use numpy array processing rather than for loops.
- Deleting the 3D cubes as soon as possible after they've been used.

After these changes, I found the KGO had changed for the acceptance test. On investigation, I think the original code had a bug as these points had a negative vertical index which should have yielded a zero hail size, but did not. I therefore think that this PR has fixed a bug. I'm not sure why it doesn't show up in the unit tests though, possibly something to do with the array flattening, although this seems unlikely.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

MoseleyS commented:
Processing of global model data still requires 40GiB of memory (and nearly 5 minutes on SPICE). I believe the vast majority of this is the extraction of the pressure of the -5C isotherm and the relative humidity of the same isotherm. While these have to run in serial, I believe I can reduce the memory requirement by refactoring these steps into individual CLIs - so I will try this.

mspelman07 reviewed (changes requested):
In general it all looks fine with just a few minor point.

MoseleyS commented on improver/psychrometric_calculations/hail_size.py:
```diff
@@ -678,9 +682,17 @@ def process(
             temperature_on_pressure
         )
 
+        attributes = {}
```
Removed

MoseleyS commented on improver/psychrometric_calculations/hail_size.py:
```diff
@@ -485,53 +498,49 @@ def get_hail_size(
-    def updated_hail_size(self, hail_size: int, wet_bulb_height: float) -> np.int8:
+    def updated_hail_size(
+        self, hail_size: np.array, wet_bulb_height: np.array
+    ) -> np.int8:
```
Updated

MoseleyS commented on improver/psychrometric_calculations/hail_size.py:
```diff
@@ -485,53 +498,49 @@ def get_hail_size(
+        ).filled(0)
+        horizontal_clipped = np.ma.where(
+            (vertical_rounded >= 0) & (horizontal_rounded >= 0), horizontal_clipped, 0
+        ).filled(0)
```
I haven't done anything with this. We need a test that has NaN temperature and RH values in the bottom one or two pressure levels to simulate the sort of data we get from StaGE (these levels can fall below the model orography, so are undefined). It may be that this "filled" method is needed to keep these cases happy.

MoseleyS replied:
I think I've fixed this now

mspelman07 reviewed (changes requested):
Thanks for working on this Stephen. Ive left some comments for you however Codecov is very unhappy with the new class and it seems to think quite a lot of lines aren't covered by tests. Im not sure I agree with it but its worth taking a look at.

MoseleyS commented on improver/psychrometric_calculations/hail_size.py:
```diff
@@ -42,19 +43,221 @@
+                variable.append(var.data[indices, lat, long])
+
+        variable_cube = pressure.copy(data=variable)
+        return variable_cube.data
```
Changed

MoseleyS commented on improver_tests/psychrometric_calculations/hail_size/test_hail_size.py:
```diff
@@ -176,74 +158,49 @@ def cube_shape_check(hail_cube):
     expected,
 ):
     """Tests the hail_size plugin with values for ccl temperature, ccl pressure,
     wet_bulb_freezing_height and relative humidity to check for expected result.
```
Done

MoseleyS commented on improver/cli/cloud_condensation_level.py:
```diff
@@ -64,15 +64,23 @@ def process(*cubes: cli.inputcube, model_id_attr: str = None):
     )
+    from improver.utilities.flatten import flatten
+
+    cubes = flatten(cubes)
```
Done

MoseleyS commented on improver/psychrometric_calculations/hail_size.py:
```diff
@@ -42,19 +43,221 @@
+        for press, var, grid in zip(press_slices, var_slices, grid_slices):
+            counter += 1
+            pressure_diff = abs(grid - press.data)
+            indices = np.nanargmin(pressure_diff, axis=0)
```
I've deleted this function as it isn't used.

MoseleyS commented on improver/psychrometric_calculations/hail_size.py:
```diff
@@ -42,19 +43,221 @@
 from improver.utilities.cube_manipulation import enforce_coordinate_ordering
 
 
+class ExtractPressureLevel(BasePlugin):
```
I suppose this was what I intended when I refactored it, but I didn't get that far. Now moved to `improver.utilities.cube_extraction`.

MoseleyS commented on improver/psychrometric_calculations/hail_size.py:
```diff
@@ -42,19 +43,221 @@
+        return pressure_cube
+
+    def extract_level_at_pressure(self, cube: Cube, pressure: Cube) -> Cube:
+        """Extract relative humidity at pressure of the environment at 268.15K
```
This function has been deleted.

MoseleyS commented:
> Thanks for working on this Stephen. Ive left some comments for you however Codecov is very unhappy with the new class and it seems to think quite a lot of lines aren't covered by tests. Im not sure I agree with it but its worth taking a look at.

There were a few methods in there that were no longer being used (a consequence of removing the relative humidity on pressure levels from the hail size plugin). I've removed those methods, moved the `ExtractPressureLevel` plugin and added tests for it to show how it works.

mspelman07 reviewed (changes requested):
One minor doc string update needed but otherwise looks good

MoseleyS commented on improver_tests/utilities/cube_extraction/test_ExtractPressureLevel.py:
```diff
@@ -0,0 +1,135 @@
+
+
+def metadata_check(pressure_slice_cube: Cube, value: float, units: str):
+    """Checks the hail cube produced by plugin has the expected metadata."""
```
Fixed this, and the name of the test function too.

MoseleyS commented on improver/utilities/cube_extraction.py:
```diff
@@ -308,3 +309,123 @@ def thin_cube(cube: Cube, thinning_dict: Dict[str, int]) -> Cube:
+            one_slice = variable_on_pressure_levels.slices_over(
+                (["realization"])
+            ).next()
+            has_r_coord = True
```
I could, but then I'd have to work out how to slice the array along an nth dimension.

MoseleyS commented on improver/utilities/cube_extraction.py:
```diff
@@ -308,3 +309,123 @@ def thin_cube(cube: Cube, thinning_dict: Dict[str, int]) -> Cube:
+
+        self.value_of_pressure_level = value_of_pressure_level
+
+    def process(self, cube: Cube) -> Cube:
```
No reason why it has to be though.
I've bowed to the convention and moved it to the bottom.

MoseleyS commented on improver/utilities/cube_extraction.py:
```diff
@@ -308,3 +309,123 @@ def thin_cube(cube: Cube, thinning_dict: Dict[str, int]) -> Cube:
+                result_data = interp_data
+        result_data = np.ma.masked_invalid(result_data)
+
+        pressure_cube = next(
+            variable_on_pressure_levels.slices_over(["pressure"])
+        ).copy(result_data)
+        pressure_cube.rename(
+            "pressure_of_atmosphere_at_"
+            f"{self.value_of_pressure_level}{variable_on_pressure_levels.units}"
+        )
+        pressure_cube.units = variable_on_pressure_levels.coord("pressure").units
+        pressure_cube.remove_coord("pressure")
```
Good point. Originally, this class had a second use case too, which I've now ditched.
Both suggestions done.

MoseleyS commented on improver/psychrometric_calculations/hail_size.py:
```diff
@@ -551,34 +405,40 @@ def hail_size_data(
-                Cube of humidity mixing ratio at the pressure of the environment at 268.15K
-            wet_bulb_zero
+            humidity_mixing_ratio_at_ccl:
+                Array of humidity mixing ratio at the pressure of the environment at the CCL
```
I don't think we're interested in the environment humidity at this point. We are interested in the behaviour of surface-driven convection. The CCL is the point at which air lifted from the surface reaches saturation point. The specific humidity of the CCL is therefore the same as at the surface (but I don't have access to that in this plugin). As I know it will be the saturation point, I can rederive the saturation humidity at the CCL and use this to drive saturated adiabatic ascent of surface air above the CCL to the point I need (-5C isotherm).

MoseleyS commented on improver/cli/cloud_condensation_level.py:
```diff
@@ -64,15 +64,23 @@ def process(*cubes: cli.inputcube, model_id_attr: str = None):
+
+    cubes = flatten(cubes)
+    (temperature, pressure, humidity,) = CubeList(cubes).extract(
+        ["air_temperature", "air_pressure", "relative_humidity"]
```
Done

MoseleyS commented:
Acceptance test data merged
